content/articles/protobuf-instead-of-json-in-frontend.md

```diff
@@ -0,0 +1,107 @@
+You also need to serialize them before sending, using the `SerializeToString()` method in our app.  
+
+### Client-side
+The client is written in Angular purely as an example and because ng-repeat was exactly what I wanted here. The code is a quick prototype so don't expect it to be idiomatic.  
+It uses [ProtoBuf.js](https://github.com/dcodeIO/ProtoBuf.js) as a Protobuf client.  
+The first thing we do is a basic GET to get all our contacts which is trivial since you can just dump the response data into our scope variable. On the other hand, we need to handle Protobuf deserialization since looking at our network tab we receive binary data looking like:
+
+```
+.
+MarlonSheets
+7771 Eastern Avenue11239
+,
+HarleyMalloy60139"
+(712)614-9303
+Y
+Mack  Hendricks3
+9573 Kimridge Cove
+7490 Orchard Hill Cove11239"
+(802)466-7004
+
+MadelineBowen87023
+0
+WillaSadler
+693 Burlington Parkway43824
+T
+```
+By default, this is more compact than JSON by quite a bit but once I turned on GZIP, the difference became much smaller: 851B for ProtoBuf and 942B for JSON.  
+
+In the example I copied the .proto file in the static directory and created our [models](https://github.com/WeAreWizards/protojson/blob/master/static/main.js#L12-L15) following the ProtoBuf.js tutorial.  
+Now the only two things left to do for our GET is to add a `responseType: 'arraybuffer'` in our HTTP requests (which would be done in the config for a proper angular app) and decode the data we received using `MyModel.decode(data)`.  
+Creating an instance looks like a class in javascript as well:
+
+```js
+var contact = new Contact({
+  first_name: _contact['firstName'],
+  last_name: _contact['lastName'],
+});
+```
+Note that supplying an incorrect param such as an unknown field will result in a runtime error.  
+
+POSTing is a bit more complex as Angular tries to be too clever and forces us to do some changes in the HTTP request:
+
+```js
+// The transformRequest is needed
+var req = {
+  method: 'POST',
+  url: '/api/contacts',
+  responseType: 'arraybuffer',
+  transformRequest: function(r) { return r;},
+  data: contact.toArrayBuffer(),
+  headers: {
+    'Content-Type': 'binary/octet-stream'
+  }
+};
+```
+
+## Conclusion
+While I see the need for Protobuf and Thrift for services communication, I don't really see the point of using it instead of JSON for the frontend. 
+First, the network tab becomes pretty much unusable and I use it very often.  
+Since Protobuf doesn't use setters, you can screw up your object and it will only throw an error much later, but earlier than with vanilla javascript. Using [Typescript](http://www.typescriptlang.org/) or [Flow](http://flowtype.org/) with good definitions would solve many issues at compile time.  
```

Comment:
ok, in that case maybe we can be more explicit and say "Since Protobuf doesn't use setters, you can set fields with data of the wrong type on your objects"